### PR TITLE
Null check GL hook driver before calling CheckImplicitThread

### PR DESCRIPTION
**Pull request by cmartel**
Explanation for the issue in https://github.com/baldurk/renderdoc/issues/1272#issuecomment-635641208

### CODE REVIEW

**Comment by baldurk** [comment]
Quoting the explanation here. Splitting the discussion between two issues makes things harder to track.

> It looks like commit [feafbde](https://github.com/baldurk/renderdoc/commit/feafbde8bca1e69d00fea18f31909f637a1a3fb3) introduced a crash on macOS GL capture replay. Reverting that commit works around the crash.
> 
> While creating the NSOpenGLContext, the `[context update]` call ends up calling glGetIntegerv internally. But because glGetIntegerv is replaced with the interposed function, it calls the `CheckImplicitThread` before the CGL driver is fully initialized, and crashes on a null pointer execution.
> 
> Call stack for the crashing thread looks like:
> 
> ```
> 0   librenderdoc.dylib            	0x000000010cb91ad0 WrappedOpenGL::CheckImplicitThread() + 16 (gl_driver.cpp:998)
> 1   librenderdoc.dylib            	0x000000010cf28328 glGetIntegerv_renderdoc_hooked(RDCGLenum, int*) + 56 (gl_hooks.cpp:151)
> 2   librenderdoc.dylib            	0x000000010cf284eb interposed_glGetIntegerv + 27 (gl_hooks.cpp:151)
> 3   com.apple.AppKit              	0x00007fff391415a1 -[NSOpenGLContext(NSOpenGLContextInternal) _updateOpenGLViewport] + 91
> 4   com.apple.AppKit              	0x00007fff39141c1e -[NSOpenGLContext update] + 408
> 5   librenderdoc.dylib            	0x000000010d0d625a NSGL_createContext + 634 (cgl_platform.mm:69)
> 6   librenderdoc.dylib            	0x000000010d0d850e CGLPlatform::MakeOutputWindow(WindowingData, bool, GLWindowingData) + 814 (cgl_platform.cpp:196)
> 7   librenderdoc.dylib            	0x000000010cb6ec5a GLReplay::MakeOutputWindow(WindowingData, bool) + 250 (gl_outputwindow.cpp:215)
> 8   librenderdoc.dylib            	0x000000010e16e301 ReplayOutput::ReplayOutput(ReplayController*, WindowingData, ReplayOutputType) + 545 (replay_output.cpp:116)
> 9   librenderdoc.dylib            	0x000000010e16e580 ReplayOutput::ReplayOutput(ReplayController*, WindowingData, ReplayOutputType) + 64 (replay_output.cpp:127)
> 10  librenderdoc.dylib            	0x000000010e189fa6 ReplayController::CreateOutput(WindowingData, ReplayOutputType) + 854 (replay_controller.cpp:1798)
> 11  org.renderdoc.qrenderdoc      	0x000000010b196898 BufferViewer::OnCaptureLoaded()::$_22::operator()(IReplayController*) const + 104 (BufferViewer.cpp:2303)
> 12  org.renderdoc.qrenderdoc      	0x000000010b196820 void std::__1::__invoke_void_return_wrapper<void>::__call<BufferViewer::OnCaptureLoaded()::$_22&, IReplayController*>(BufferViewer::OnCaptureLoaded()::$_22&&&, IReplayController*&&) + 80 (__functional_base:350)
> 13  org.renderdoc.qrenderdoc      	0x000000010b196719 std::__1::__function::__func<BufferViewer::OnCaptureLoaded()::$_22, std::__1::allocator<BufferViewer::OnCaptureLoaded()::$_22>, void (IReplayController*)>::operator()(IReplayController*&&) + 57 (functional:1562)
> 14  org.renderdoc.qrenderdoc      	0x000000010acab384 std::__1::function<void (IReplayController*)>::operator()(IReplayController*) const + 132 (functional:1913)
> 15  org.renderdoc.qrenderdoc      	0x000000010acaaf42 ReplayManager::run(int, QString const&, ReplayOptions const&, std::__1::function<void (float)>) + 1858 (ReplayManager.cpp:482)
> 16  org.renderdoc.qrenderdoc      	0x000000010acaf283 ReplayManager::OpenCapture(QString const&, ReplayOptions const&, std::__1::function<void (float)>)::$_0::operator()() const + 115 (ReplayManager.cpp:55)
> 17  org.renderdoc.qrenderdoc      	0x000000010acaf1fd void std::__1::__invoke_void_return_wrapper<void>::__call<ReplayManager::OpenCapture(QString const&, ReplayOptions const&, std::__1::function<void (float)>)::$_0&>(ReplayManager::OpenCapture(QString const&, ReplayOptions const&, std::__1::function<void (float)>)::$_0&&&) + 45 (__functional_base:350)
> 18  org.renderdoc.qrenderdoc      	0x000000010acaeda9 std::__1::__function::__func<ReplayManager::OpenCapture(QString const&, ReplayOptions const&, std::__1::function<void (float)>)::$_0, std::__1::allocator<ReplayManager::OpenCapture(QString const&, ReplayOptions const&, std::__1::function<void (float)>)::$_0>, void ()>::operator()() + 41 (functional:1562)
> 19  org.renderdoc.qrenderdoc      	0x000000010acac636 std::__1::function<void ()>::operator()() const + 102 (functional:1913)
> 20  org.renderdoc.qrenderdoc      	0x000000010acabb94 LambdaThread::process() + 68 (QRDUtils.h:339)
> 21  org.renderdoc.qrenderdoc      	0x000000010acac5ae QtPrivate::FunctorCall<QtPrivate::IndexesList<>, QtPrivate::List<>, void, void (LambdaThread::*)()>::call(void (LambdaThread::*)(), LambdaThread*, void**) + 126 (qobjectdefs_impl.h:152)
> 22  org.renderdoc.qrenderdoc      	0x000000010acac503 void QtPrivate::FunctionPointer<void (LambdaThread::*)()>::call<QtPrivate::List<>, void>(void (LambdaThread::*)(), LambdaThread*, void**) + 99 (qobjectdefs_impl.h:186)
> 23  org.renderdoc.qrenderdoc      	0x000000010acac416 QtPrivate::QSlotObject<void (LambdaThread::*)(), QtPrivate::List<>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) + 166 (qobjectdefs_impl.h:415)
> 24  org.qt-project.QtCore         	0x0000000110658d0b QMetaObject::activate(QObject*, int, int, void**) + 2219
> 25  org.qt-project.QtCore         	0x0000000110460868 QThread::started(QThread::QPrivateSignal) + 56
> 26  org.qt-project.QtCore         	0x0000000110461059 QThreadPrivate::start(void*) + 297
> 27  libsystem_pthread.dylib       	0x00007fff63b29661 _pthread_body + 340
> 28  libsystem_pthread.dylib       	0x00007fff63b2950d _pthread_start + 377
> 29  libsystem_pthread.dylib       	0x00007fff63b28bf9 thread_start + 13
> ```
> 
> Null checking `glhook.driver` before calling `CheckImplicitThread` on it also works as a naive fix, but it might be a bit too much so. Without a full understanding of the thread switching issue that required the addition of that context tracking functionality, I can't be certain it won't break anything... But as far as I can tell, the only time when `glhook.driver` would be unexpectedly null is before/during context creation, and I should hope that the CGL implementation doesn't change the context in the middle of this.

The problem is that this fix doesn't look like it addresses the core problem. Checking `glhook.driver` here may fix an immediate crash but the hook will then go on to call through `glhook.driver` for the real function anyway after that. There's a bit of macro chasing to see this, but if you look at where `SCOPED_GLCALL` is actually used in `gl_dispatch_table_defs.h` you'll see that `glhook.driver` is used immediately after. If this wasn't crashing before my guess is that it was getting lucky and only calling `glGetIntegerv` and only with enums that don't require the driver to access `this`, so it managed to survive and call into the real function `GL.glGetIntegerv`.

The root problem here is that the hooks are not expected to be called at all on replay and they aren't set up for it at all, the `glhook.driver` pointer will never be set on replay. On linux and mac though the hooking is unconditional when `librenderdoc.so` is loaded, so it's unavoidable that any code on replay calling these functions will call into the hook. On linux this doesn't seem to happen but I guess it does on mac.

I think the solution will also have to include the kind of protection that's already in the `CGL` and `GLX` functions, right at the start if we're replaying we just forward the call straight on to the real function:

```
  if(RenderDoc::Inst().IsReplayApp())
  {
    if(!CGL.CGLCreateContext)
      CGL.PopulateForReplay();

    return CGL.CGLCreateContext(pix, share, ctx);
  }
```

The best place for this is probably the `UNINIT_CALL` macro.

Within `gl_hooks.cpp` the populate function will be a bit more complex. It will need to check either `GetGLPlatform()` or `GetEGLPlatform()` depending on GL/EGL support (in that order I think), call `GLPlatform::PopulateForReplay()` then use `GLPlatform::GetReplayFunction()` to pass to `GL.PopulateWithCallback`. That should then populate the onward GL functions to call instead of the driver hook.

File: renderdoc/driver/gl/gl_hooks.cpp [hunk 1]
```diff
@@ -87,18 +87,23 @@ int ScopedPrinter::depth = 0;
...
 #else
 
-#define SCOPED_GLCALL(funcname)    \
-  SCOPED_LOCK(glLock);             \
-  gl_CurChunk = GLChunk::funcname; \
-  glhook.driver->CheckImplicitThread();
+#define SCOPED_GLCALL(funcname)           \
+  SCOPED_LOCK(glLock);                    \
+  gl_CurChunk = GLChunk::funcname;        \
+  {                                       \
```
Comment (baldurk) [inline renderdoc/driver/gl/gl_hooks.cpp:105]: There's no check on `glhook.driver` here which means release builds will still try to dereference it.

**Comment by cmartel** [comment]
Fair point regarding the function call not getting forwarded from the hook--I was fixated on resolving the crash and because the function not getting forwarded seemed asymptomatic, I didn't think to wonder why it's happening. I'll adapt the fix accordingly and test a bit.

**Comment by baldurk** [comment]
I think the function call _is_ getting forwarded - that's what worries me. It's calling `glhook.driver->glGetIntegerv` with `glhook.driver == NULL`, but calling that might not crash if you only call e.g. `glGetIntegerv(GL_MAJOR_VERSION)` and it doesn't try to access `this` inside the call.

**Comment by cmartel** [comment]
Your initial hunch about writing a workaround in UNINIT_CALL might have been spot on. I manually expanded the `glGetIntegerv` hook to get a better understanding of what's happening and to be able to debug whether it is forwarding the hook or not, and it seems like UNINIT_CALL is working as intended already.

It looks like GL hooking is disabled (as it sounds like it should be, for replaying). When stepping through the manually expanded glGetIntegerv call, `glhook.enabled` is false, so it is not actually calling the hooked version, but just directly calling the GL driver's implementation through UNINIT_CALL.

With that in mind, perhaps a better version of this crash fix would be to avoid calling CheckImplicitThread if `glhook.enabled` is false, rather than null checking.

**Comment by baldurk** [comment]
Ah interesting, that wasn't at all the intention of `UNINIT_CALL` in the first place but it seems like it sort of works. I'm a little shakey about relying on that coincidence but it's probably as good a fix. 

That does still mean though that there's a potential problem on linux where `UNINIT_CALL` isn't used, since linux has the same issue where those hooks could get called on replay by any statically linked code that doesn't use dlsym or glXGetProcAddress.

**Comment by cmartel** [comment]
True. I can't vouch for Linux, but I can at least confirm that the workaround of checking `glhook.enabled` works correctly. If that's good enough for now, great--if you'd rather do a fully fledged check of `RenderDoc::Inst().IsReplayApp()` as initially suggested, I can revisit this fix later, for now I have to get back to fixing the issue I was using macOS RenderDoc to diagnose in the first place. :)

**Review by baldurk — APPROVED** [review]
The reason I bring it up is bluntly because linux is a platform I actually support and so I care about bugs there :smile:. Mac fixes are of course welcome but not as important as ensuring the supported platforms continue to function.

I think this change is fine for now though, I'll either make a note to look into linux's setup at a later date or see what happens if someone encounters a problem there.

**Pull request merged**